lib/logstash/outputs/mongodb.rb

```diff
@@ -99,20 +112,24 @@ def receive(event)
       else
         @db[event.sprintf(@collection)].insert_one(document)
       end
-
     rescue => e
-      @logger.warn("Failed to send event to MongoDB", :event => event, :exception => e,
-                   :backtrace => e.backtrace)
       if e.message =~ /^E11000/
-          # On a duplicate key error, skip the insert.
-          # We could check if the duplicate key err is the _id key
-          # and generate a new primary key.
-          # If the duplicate key error is on another field, we have no way
-          # to fix the issue.
+        # On a duplicate key error, skip the insert.
+        # We could check if the duplicate key err is the _id key
+        # and generate a new primary key.
+        # If the duplicate key error is on another field, we have no way
+        # to fix the issue.
+        @logger.warn("Skipping insert because of a duplicate key error", :event => event, :exception => e)
       else
-        sleep @retry_delay
+        @logger.warn("Failed to send event to MongoDB, retrying in #{@retry_delay.to_s} seconds", :event => event, :exception => e)
+        sleep(@retry_delay)
         retry
       end
     end
-  end # def receive
-end # class LogStash::Outputs::Mongodb
+  end
+
+  def close
+    @closed.make_true
+    @bulk_thread.join
```

Comment:
we may also want to send [`Thread#wakeup`](https://ruby-doc.org/core-2.3.0/Thread.html#method-i-wakeup) to our `@bulk_thread` to prevent it from sleeping off the remainder of its interval unnecessarily.

Reply:
good idea.

Reply:
done